### ISSUE TITLE
Concept of SharedComponents across multiple entities

### ISSUE THREAD

**Issue by NearAutomata** [skypjack/entt#151]
I just recently stumbled upon this gem of yours and I am really glad about your design principles. Currently I am attempting to write custom wrappers around EnTT to formalize its use in my context using things such as Worlds, EntityManagers, ComponentSystems, EntityArchetypes and Chunks.

However I noticed you are using rather generic Components and seemingly have no concept of what I call SharedComponents. Assume a large amount of entities that all have the very same Component attached to it, such as a MeshRendererComponent or maybe an AgentBehaviourComponent for AI-related things. Something that all of those entities have in common. Instead of having N components attached to N entities that all hold the very same data it would be nice to have variant where a SharedComponent is only stored once and letting all entities point to this single instance of the SharedComponent.

Is this a feature you already have implemented (which I just haven't seen at all) or something that you might consider adding to EnTT?

**Comment by skypjack** [comment]
>I just recently stumbled upon this gem of yours and I am really glad about your design principles.

Thank you. Do not forget to star the project if you like it and use it! :+1:

>Instead of having N components attached to N entities that all hold the very same data it would be nice to have variant where a SharedComponent is only stored once and letting all entities point to this single instance of the SharedComponent

Ok, let's consider this. I'll try to explain you why you don't need it.

In order to have N entities that point to the same shared component, you need at least N pointers of type `T *` and one or more instances of `T` stored somewhere else. Right?  
So far, so good. In C++, if you define a class with only a public data member of type `T *`, it will have the size of its data member. So, call this class `C` and assign an instance of this class to each entity, you occupied exactly the same memory you would have occupied with N pointers of type `T *`, with the sole difference that you called them components and you put them in an heterogeneous container that treats them as any other type. Moreover, you still have your instances of `T` stored somewhere else.

Do you spot the difference? No, because there is not.

I did it recently. I have a resource cache that returns a tiny handle (for simplicity, consider it a pointer) to a 2D sprite that can be shared between different entities.  
If you define your component as:
```
class Sprite {
    Handle handle;
};
```
You have exactly what you're looking for.

It's called flyweight, it's a well known pattern, no matter how Unity calls it and if they hide it behind a graceful interface.  

There is no way you can occupy less then what you'd occupy by putting a pointer in a component and by assigning instances of that component to entities.
Moreover, there is no way you can have a performance boost using a built-in solution, because you cannot get rid of the indirection.

So, to sum up, you're asking to integrate in `EnTT` something that already exists, to do something that is already possible, just because someone called it differently (so that it looked smarter than a design pattern everybody knows about probably). :-)

Did I convince you?

**Comment by NearAutomata** [comment]
Frankly, I was wearing mindcuffs and thus neglected the option to just store a pointer. In Unity there are various high performance mechanisms that lay out the data sequentially in the memory and parallelize systems using threads as much as possible. One caveat is that you can only use blittable types and pointer are neither directly available in C# nor considered blittable unless you do some shenannigans.

With that in mind I just haven't really thought about using pointers. Thanks for the very thorough explanation 👍

**Comment by skypjack** [comment]
You're welcome. Thanks for using `EnTT`.
